### ISSUE TITLE
Implement streaming support for all backends

### ISSUE THREAD

**Issue by pavanky** [arrayfire/arrayfire#113]
Put the C functions in af/device.h

Some options to consider:
- **OPTION 1**: Use af_set_stream like af_set_device. i.e. all functions after calling af_set_stream would use the same stream.
- **OPTION 2**: Write new functions that take in the stream parameter. For example af_sum_async(&out, in, dim, stream) (or something such) vs af_sum(&out, in, dim)

**Comment by munnybearz** [comment]
+1 for Option 2

**Comment by WilliamTambellini** [comment]
Is nt this one still legitimate, at least for afcuda ?

**Comment by umar456** [comment]
Yes, This is related to streaming data from a source on the host to the device efficiently. This is different from CUDA streams although they will be used here.

**Comment by ipostr08** [comment]
In the documentation, I don't see anything like setStream or an extra parameter for a stream in functions. Was this idea abandoned?

**Comment by 9prady9** [comment]
@ipostr08 As of now, users can only fetch ArrayFire's CUDA stream. The support to set a user created CUDA stream is not yet available. No this feature is not abandoned. We are adding things (Events in 3.7 release - helps with streams) progressively, in smaller increments. Unfortunately, we don't have a ETA on this feature right now but we will update this issue with updates as we move forward.

**Comment by WilliamTambellini** [comment]
Seen at GDC 2020:
![HowtoBetterUtilizeGPU](https://user-images.githubusercontent.com/109458/77782419-81f20480-7014-11ea-8fae-8ad3970c1781.jpg)

@umar456  any news about stream support in AFcuda ?

**Comment by umar456** [comment]
We have implemented events which should improve support for streams. The main problem is the memory manager and its current implementation and how the streams are associated with each device. We will need to create a context to enable this sort of behavior.

**Comment by NOBLES5E** [comment]
> We have implemented events which should improve support for streams. The main problem is the memory manager and its current implementation and how the streams are associated with each device. We will need to create a context to enable this sort of behavior.

Can we get the internal stream that arrayfire uses?

**Comment by 9prady9** [comment]
> > We have implemented events which should improve support for streams. The main problem is the memory manager and its current implementation and how the streams are associated with each device. We will need to create a context to enable this sort of behavior.
> 
> Can we get the internal stream that arrayfire uses?

https://arrayfire.org/docs/group__cuda__mat.htm#gaec1dc4c2aa935dc61889f23248c8450d

**Comment by NatanBiesmans** [comment]
Any news on this? 
My throughput is terrible because my program spends most of the time sending data to and from the GPU. It would be great to be able to do operations while the next data is being send to the GPU.
